File: leetcode/[0046]permutations.ts
// https://leetcode.com/problems/permutations/
function permute(nums: number[]): number[][] {
    return solution2(nums);
}



// remove duplicates from all cases: 4 ms
function solution1(nums: number[]): number[][] {
    return [...permute1(nums, nums.length)];
}
function* permute1<T>(src: T[], k: number): Generator<T[]> {
    k = Math.max(Math.min(src.length, k), 0);
    const seeds = Array.from({ length: src.length }, (_, i) => i);

    let pools = [[]] as number[][];
    for (let i = 0; i < k; i++) {
        const tmp = [];
        for (const pool of pools) {
            for (const x of seeds) {

                if (pool.indexOf(x) !== -1) continue;

                tmp.push([...pool, x]);
            }
        }
        pools = tmp;
    }

    for (const pool of pools) {
        yield pool.map((i) => src[i]);
    }
}



// assemble permutation: 3 ms
function solution2(nums: number[]): number[][] {
    return [...permute2(nums, nums.length)];
}
function* permute2<T>(src: T[], k: number): Generator<T[]> {
    k = Math.max(Math.min(src.length, k), 0);
    const n = src.length;
    const pool = Array.from({ length: k }, (_, i) => i);

    let idx = k - 1;
    while (true) {
        yield pool.map((i) => src[i]);

        while (true) {
            pool[idx] += 1;

            if (pool[idx] === n) {
                idx -= 1;
                if (idx < 0) return;

                continue;
            }

            if (pool.slice(0, idx).indexOf(pool[idx]) !== -1) continue;

            let i = 0;
            for (let j = idx + 1; j < k; j++) {
                while (pool.slice(0, idx + 1).indexOf(i) !== -1) {
                    i += 1;
                }
                pool[j] = i;
                i += 1;
            }

            idx = k - 1;
            break;
        }
    }
}

File: leetcode/[0046]permutations.rs
// https://leetcode.com/problems/permutations/
impl Solution {
    pub fn permute(nums: Vec<i32>) -> Vec<Vec<i32>> {
        return solution2(nums);
    }
}



// assemble permutations: 0 ms
fn solution1(nums: Vec<i32>) -> Vec<Vec<i32>> {
    let k = nums.len();
    
    return permute1(nums, k);
}
fn permute1<T: Clone>(src: Vec<T>, k: usize) -> Vec<Vec<T>> {
    use std::cmp::min;

    let k = min(k, src.len());
    let pool = Vec::from_iter(0..src.len());

    let mut a = Vec::<Vec<usize>>::new();
    a.push(vec![]);

    for i in 0..k {
        let mut tmp = Vec::<Vec<usize>>::new();

        for seed in a.iter() {
            for &x in pool.iter() {

                if seed.contains(&x) {
                    continue;
                }

                let mut t = seed.clone();
                t.push(x);
                tmp.push(t);
            }
        }

        a = tmp;
    }

    let a = a
        .into_iter()
        .map(|v| v.into_iter().map(|i| src[i].clone()).collect::<Vec<T>>())
        .collect::<Vec<Vec<T>>>();
    
    return a;
}



// assemble permutations2: 0 ms
fn solution2(nums: Vec<i32>) -> Vec<Vec<i32>> {
    let k = nums.len();

    return permute2(nums, k);
}
fn permute2<T: Clone>(src: Vec<T>, k: usize) -> Vec<Vec<T>> {
    use std::cmp::min;

    let k = min(k, src.len());
    let mut a = Vec::<Vec<usize>>::new();
    
    let mut seed = (0..k).collect::<Vec<usize>>();
    let mut idx = k - 1;

    'outer:
    loop {
        a.push(seed.clone());

        loop {
            seed[idx] += 1;

            if seed[idx] == src.len() {
                idx -= 1;

                if (idx as isize) < 0 {
                    break 'outer;
                }

                continue;
            }

            if seed[0..idx].contains(&seed[idx]) {
                continue;
            }

            let mut n = 0;
            for j in (idx + 1)..k {
                while seed[0..j].contains(&n) {
                    n += 1;
                }

                seed[j] = n;
            }

            idx = k - 1;
            break;
        }
    }

    let a = a
        .into_iter()
        .map(|v| v.into_iter().map(|i| src[i].clone()).collect::<Vec<T>>())
        .collect::<Vec<Vec<T>>>();

    return a;
}
